### ISSUE TITLE
SFT中 loss mask与最终loss没有对齐，会mask掉eos_token的loss

### ISSUE THREAD

**Issue by Fyweven** [volcengine/verl#3562]
### System Info

在fsdp_sft_trainer.py中，最新的代码，loss_mask丢弃第一个token，向左shift了

而loss计算时，丢弃的是最后一个token
eos_token的loss_mask本应该是1，向左shift后，变成了0

导致SFT训练后的模型，经常在回复的结尾陷入循环回复，无法终止

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

loss_mask = batch.pop("loss_mask")[:, 1:].reshape(-1).to(self.device_name) 
full_loss = full_loss.squeeze(-1)[:, :-1]  # Remove last token's loss
full_loss = full_loss.reshape(-1)
loss_mask = loss_mask.to(full_loss.device)
loss = full_loss * loss_mask

### Expected behavior

loss_mask = batch.pop("loss_mask")[:, :-1].reshape(-1).to(self.device_name) 
full_loss = full_loss.squeeze(-1)[:, :-1]  # Remove last token's loss
full_loss = full_loss.reshape(-1)
loss_mask = loss_mask.to(full_loss.device)
loss = full_loss * loss_mask

**Comment by Linzwcs** [comment]
The same issue. I think it is indeed a bug.

code
```python
loss_mask = batch.pop("loss_mask")[:, :-1].reshape(-1).to(self.device_name)
```
should be correct.

**Comment by Caelyn-gao** [comment]
I also noticed the change. The reason for the change can be found in: https://github.com/volcengine/verl/pull/3287. I think it was fixed for multiple-turn dataset.

And I agree "loss_mask = batch.pop("loss_mask")[:, 1:].reshape(-1).to(self.device_name)" will cause endless repetition for single-turn dataset.

**Comment by Asher-a-Fisher** [comment]
Same question. Why mask the loss of the last token (usually eos)?

**Comment by kkkaiaiai** [comment]
似乎应该将下面这行代码去除，这样最后的loss mask是正确的，对于response的第一个token和最后一个token（EOS）的预测都会计算相应的loss
https://github.com/volcengine/verl/blob/504696245b39dd22162579b36303706ee61a731a/verl/utils/dataset/sft_dataset.py#L197

**Comment by lostella** [comment]
I agree this looks like a bug: the -1 should be removed [here](https://github.com/volcengine/verl/blob/504696245b39dd22162579b36303706ee61a731a/verl/utils/dataset/sft_dataset.py#L195) and [here](https://github.com/volcengine/verl/blob/504696245b39dd22162579b36303706ee61a731a/verl/utils/dataset/sft_dataset.py#L197).

**Comment by lsy641** [comment]
yes, it is a bug

**Comment by Kirrito-k423** [comment]
中文：感谢报告。这个问题本身是有效的：旧的 `SFTDataset`/`fsdp_sft_trainer.py` 路径里，loss mask 与 shifted labels 的对齐会导致单轮 SFT 场景下 EOS token 的 loss 被错误 mask 掉。后续相关 PR #5362 也复现并解释了这一点。

中文：不过当前 main 分支已经不再使用这条旧路径。PR #5382 已合入并移除了 `verl/utils/dataset/sft_dataset.py` 和 `verl/trainer/fsdp_sft_trainer.py`，当前 SFT 默认走 `MultiTurnSFTDataset` + `sft_loss`：dataset 侧对 assistant token 保留 loss mask，loss 函数侧统一左移 mask 来对齐 log_prob。因此原 issue 指向的旧 bug 在当前代码路径中已经解决，可以关闭。

English: Thanks for the report. The original issue was valid: in the old `SFTDataset`/`fsdp_sft_trainer.py` path, the loss mask was misaligned with the shifted labels, so EOS loss could be masked out for single-turn SFT. PR #5362 also reproduced and explained this behavior.

English: The affected path no longer exists on current `main`. PR #5382 has been merged and removed both `verl/utils/dataset/sft_dataset.py` and `verl/trainer/fsdp_sft_trainer.py`. Current SFT defaults to `MultiTurnSFTDataset` plus `sft_loss`: the dataset keeps assistant tokens, including EOS, in the loss mask, and the loss function performs the mask shift to align with log probabilities. So the original bug is resolved in the current code path and this issue can be closed.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
